packages/app/lib/ui/widgets/pagination_list.dart

```diff
@@ -0,0 +1,251 @@
+
+typedef PaginationBuilder<T> = Widget Function(
+  T document,
+);
+
+typedef ContainerBuilder<T> = Widget Function(
+  List<T> collection,
+);
+
+class PaginationBase<T> extends StatelessWidget {
+  final ContainerBuilder<T> builder;
+  final Paginator<T> paginator;
+
+  const PaginationBase(
+      {super.key, required this.builder, required this.paginator});
+
+  Widget _error(BuildContext context, String errorMessage) {
+    return ErrorCard(
+        message:
+            AppLocalizations.of(context)!.paginationListError(errorMessage));
+  }
+
+  Widget _loading() {
+    return Center(
+        child: Container(
+            padding: const EdgeInsets.all(30.0),
+            child: const CircularProgressIndicator(
+              color: MeliColors.black,
+            )));
+  }
+
+  Widget _emptyResult(BuildContext context) {
+    return InfoCard(
+        message: AppLocalizations.of(context)!.paginationListNoResults);
+  }
+
+  Widget _loadMore(BuildContext context, bool isLoading,
+      {required VoidCallback onLoadMore}) {
+    return Column(children: [
+      isLoading
+          ? this._loading()
+          : ElevatedButton(
+              onPressed: onLoadMore,
+              child: Text(AppLocalizations.of(context)!.paginationListLoadMore,
+                  style: const TextStyle(color: MeliColors.black)),
+            )
+    ]);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Query(
+      options: QueryOptions(document: this.paginator.nextPageQuery(null)),
+      builder: (result, {VoidCallback? refetch, FetchMore? fetchMore}) {
+        if (this.paginator.refresh == null) {
+          // Workaround to access `refetch` method from the outside
+          this.paginator.refresh = refetch;
+        }
+
+        if (result.hasException) {
+          return this._error(context, result.exception.toString());
+        }
+
+        if (result.isLoading && result.data == null) {
+          return this._loading();
+        }
+
+        final data =
+            this.paginator.parseJSON(result.data as Map<String, dynamic>);
+
+        FetchMoreOptions opts = FetchMoreOptions(
+          document: this.paginator.nextPageQuery(data.endCursor),
+          updateQuery: (previousResultData, fetchMoreResultData) {
+            return this
+                .paginator
+                .mergeResponses(previousResultData!, fetchMoreResultData!);
+          },
+        );
+
+        if (data.documents.isEmpty) {
+          return this._emptyResult(context);
+        }
+
+        return Column(
+          crossAxisAlignment: CrossAxisAlignment.stretch,
+          children: <Widget>[
+            this.builder(data.documents),
+            if (data.hasNextPage)
+              this._loadMore(context, result.isLoading, onLoadMore: () {
+                fetchMore!(opts);
+              })
+          ],
+        );
+      },
+    );
+  }
+}
+
+class SliverPaginationBase<T> extends StatelessWidget {
```

Comment:
This variant uses slivers instead of widgets, this might be a good starting point for this refactor: https://github.com/p2panda/meli/issues/77